CPAOT support for inlined PInvokes

trylek opened the pull request:
This change adds previously missing implementation of two JIT
interface methods dealing with inlined PInvokes -
getAddressOfPinvokeTarget and pInvokeMarshalingRequired.

For the MethodRequiresMarshaling check I used the same technique
JanK recommended before for IsManagedSequential - I added an extra
instrumentation to Crossgen and CPAOT to capture PInvoke methods
and the result of the marshaling check; with this change there are
0 differences in the Pri#1 CoreCLR test suite.

Thanks

Tomas

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1662,6 +1664,96 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+                return true;
+            }
+
+            // The return type size doesn't contribute to stackSize
```
Nit: Computing stack size in the `TypeRequiresMarshaling` makes the method a bit awkward to use. Are we really getting something out of it?

trylek replied:
Removed.

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1662,6 +1664,96 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+            {
+                case TypeFlags.Pointer:
+                    // TODO: custom modifiers S(NeedsCopyConstructorModifier, IsCopyConstructed)
+                    break;
```
Breaking out here will end up throwing `InvalidCastException` on line 1755 because pointers are not `DefType`.

trylek replied:
Based on your and JanK's feedback I have removed the stackSize parameter from the method so we no longer need the DefType cast.

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1662,6 +1664,96 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+                    // TODO: custom modifiers S(NeedsCopyConstructorModifier, IsCopyConstructed)
+                    break;
+
+                case TypeFlags.Enum:
```
Nit: might be more readable to have the `TypeFlags.Enum` case be just `_compilation.NodeFactory.CompilationModuleGroup.VersionsWithType(type);`.

It would simplify two of the `if` checks in the combined ValueType case right now.

trylek replied:
That is true, thanks for pointing that out, that is indeed easier to understand.

MichalStrehovsky commented on src/ILCompiler.RyuJit/src/JitInterface/CorInfoImpl.RyuJit.cs:
```diff
@@ -1541,5 +1541,30 @@ private void getGSCookie(IntPtr* pCookieVal, IntPtr** ppCookieVal)
+
+        private bool pInvokeMarshalingRequired(CORINFO_METHOD_STRUCT_* handle, CORINFO_SIG_INFO* callSiteSig)
+        {
+            // calli is covered by convertPInvokeCalliToCall
```
Can you also move `convertPInvokeCalliToCall` here? The ReadyToRun version can just throw NotImplemented. Should be unreachable.

trylek replied:
Good point, forked.

MichalStrehovsky commented on src/ILCompiler.RyuJit/src/JitInterface/CorInfoImpl.RyuJit.cs:
```diff
@@ -1541,5 +1541,30 @@ private void getGSCookie(IntPtr* pCookieVal, IntPtr** ppCookieVal)
+
+            // We could have given back the PInvoke stub IL to the JIT and let it inline it, without
+            // checking whether there is any stub required. Save the JIT from doing the inlining by checking upfront.
+            return IsPInvokeStubRequired(method);
```
Can you also move this method?

trylek replied:
Moved to CorInfoImpl.RyuJit.

jkotas commented on src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1662,6 +1664,96 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+                    return true;
+                }
+            }
+            return !FitsInU2(stackSize);
```
I do not think we need this `FitsInU2` check here.

trylek replied:
OK, thanks, that helps as it also lets me address another Michal's comment by removing the stackSize parameter from TypeRequiresMarshaling.

trylek commented on src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1662,6 +1664,96 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+
+        private bool pInvokeMarshalingRequired(CORINFO_METHOD_STRUCT_* handle, CORINFO_SIG_INFO* callSiteSig)
+        {
+            EcmaMethod method = (EcmaMethod)HandleToObject(handle);
```
I have added the null check. At this point I don't exactly understand what is supposed to happen in the null case so for now I put a NotImplementedException in the else branch.

MichalStrehovsky commented on src/ILCompiler.ReadyToRun/src/JitInterface/CorInfoImpl.ReadyToRun.cs:
```diff
@@ -1662,6 +1664,112 @@ private void expandRawHandleIntrinsic(ref CORINFO_RESOLVED_TOKEN pResolvedToken,
+                    {
+                        return true;
+                    }
+                    if (isReturnType && !type.UnderlyingType.IsPrimitive)
```
```suggestion
                    if (isReturnType && !type.IsPrimitive)
```

trylek commented:
I believe I have addressed all review comments from @MichalStrehovsky and @jkotas; please take another look when you have a chance. Thanks for the useful feedback!
